### ISSUE TITLE
Do we have to resample when performing inference ?

### ISSUE THREAD

**Issue by tomDag25** [ivadomed/model-canal-seg#33]
The question has been asked during the SCT meeting (2026-03-31) whether we should resample and reorient before inference for our nnUNet models or not.
The underlying idea is that nnUNet is supposed to already perform resampling and reorientation so is it necessary to do so ourselves ? 

On all images 🟣 is the segmentation with resampling and reorientation 🟠 is the comparison. 
To answer this question I took ten images referenced below and compared the results. Three main results came out of this experiment: 

1. **Reorientation is mandatory** ->  As you can see here the model is completely lost without any reorientation for inference and the results make no sense. 
 

![Image](https://github.com/user-attachments/assets/58617f41-a838-4034-a328-01c4c4f08a3e)

![Image](https://github.com/user-attachments/assets/b5ce604a-fb06-416b-ba48-73f03c865a2e)

 
2. **Resampling has an impact on the prediction** -> Even if the results seem pretty similar you can see some differences in the case where the image wasn't natively in 1mm iso (in that case no resampling was necessary anyway). It can be argued that it doesn't lower the performances but it does change the prediction. 

Resolution != 1mm iso 

![Image](https://github.com/user-attachments/assets/c566d782-bb89-4d9d-9413-93e4232dc4d2)

Resolution == 1mm iso 

![Image](https://github.com/user-attachments/assets/031a7d12-5545-4781-877e-f3aaf9473d8d)

3. **Resampling is not adding a significant amount of time** -> When comparing the inference time we can clearly see that inference time is mostly due to nnUNet and not the resampling. 

```
subject,version,inference_time_seconds
sub-001_STIR.nii.gz,no_resampling,16
sub-001_STIR.nii.gz,with_resampling,18
sub-008_acq-sag_T1w.nii.gz,no_resampling,17
sub-008_acq-sag_T1w.nii.gz,with_resampling,18
sub-008_acq-sag_T2w.nii.gz,no_resampling,17
sub-008_acq-sag_T2w.nii.gz,with_resampling,19
sub-073_PSIR.nii.gz,no_resampling,18
sub-073_PSIR.nii.gz,with_resampling,18
sub-095_MP2RAGE.nii.gz,no_resampling,19
sub-095_MP2RAGE.nii.gz,with_resampling,19
sub-cal092_ses-M0_acq-T1w_MTS.nii.gz,no_resampling,17
sub-cal092_ses-M0_acq-T1w_MTS.nii.gz,with_resampling,18
sub-milanFilippi100_acq-ax_T2star.nii.gz,no_resampling,16
sub-milanFilippi100_acq-ax_T2star.nii.gz,with_resampling,18
sub-milanFilippi100_acq-sag_T2w.nii.gz,no_resampling,16
sub-milanFilippi100_acq-sag_T2w.nii.gz,with_resampling,17
sub-NSPilot007_ses-01spinalcord_task-FingerTap_bold_mc2.nii.gz,no_resampling,16
sub-NSPilot007_ses-01spinalcord_task-FingerTap_bold_mc2.nii.gz,with_resampling,16
```

**Comment by NathanMolinier** [comment]
Maybe the difference you see when letting nnunet do the resampling/ when doing the resampling before giving the data to nnunet, is due to different resampling method. For example spline vs linear.

**Comment by valosekj** [comment]
Just to double-check, are you sure that nnUNet resamples to 1mm iso? If I'm not mistaken, nnUNet resamples to the median resolution.
In other words, it's easy to control the resampling when doing it manually before running nnUNet. But it's hard to control it when it's done by nnUNet as it uses median across images.

**Comment by tomDag25** [comment]
Since my training set is resampled to 1mm iso before training the median should be 1mm iso. My understanding from the plans.json is that it does resample to 1mm iso but I have to check. 
In particular this part: 
```
"spacing": [
                1.0,
                1.0,
                1.0
            ],
```



[plans.json](https://github.com/user-attachments/files/26575823/plans.json)

**Comment by valosekj** [comment]
> Since my training set is resampled to 1mm iso before training the median should be 1mm iso. My understanding from the plans.json is that it does resample to 1mm iso but I have to check. In particular this part:

Yes, I think this is expected when doing the resampling manually before calling nnUNet.

> ```
> "spacing": [
>                 1.0,
>                 1.0,
>                 1.0
>             ],
> ```

LGTM.

Maybe I should have asked my question in a different way: what would be the resolution determined by nnUNet if you didn't resample before? Likely not 1mm iso. And what would be the impact on the model performance?

**Comment by tomDag25** [comment]
Oh ok I see. Good question. I think I tried it before but I needed to clean my tassan because of space issues so I don't have the results anymore. The performance got enhanced when doing the resampling before nnUNet if I remember correctly.  
Also performing the resampling beforehand solves the issue related to the fact that nnUnet can change the axis order based on the resolution when dealing with axial vs sagittal.

**Comment by tomDag25** [comment]
I performed some quantitative analysis to compare the results when performing resampling prior to inference or not. Here are the results when the inference is performed on the 151 patients from the test set : 

| Metric | With Resampling | Without Resampling |
| :--- | :--- | :--- |
| Dice | 0.9476 ± 0.0208 | 0.9546 ± 0.0218 | 
| Time | 19.91 ± 2.69 | 18.42 ± 1.91 | 

Looking at these results **there doesn't seem to be any valid reason to perform resampling beforehand** for inference as the dice is even a bit better when doing it after.
